### ISSUE TITLE
Hyprland/workspaces improvements and further development

### ISSUE THREAD

**Issue by zjeffer** [Alexays/Waybar#2271]
I'm creating this issue so we can keep track of further development for the newly implemented hyprland/workspaces module (by @Anakael in https://github.com/Alexays/Waybar/pull/2264)

Here are some features I would personally like to see:

- [ ] Urgent class
- [ ] Focused class
- [ ] Option to only show active workspaces
- [ ] Persistent workspaces (+persistent css class)
- [ ] Multi-monitor: option to disable showing all workspaces on every output
- [ ] on-click action
- [ ] on-scroll action

There is also currently a segfault happening when rapidly switching workspaces: https://github.com/Alexays/Waybar/pull/2264#issuecomment-1615772176.

**Comment by Anakael** [comment]
Active (focused) class is already implemented.

**Comment by Anakael** [comment]
Wiki is updated:
https://github.com/Alexays/Waybar/wiki/Module:-Hyprland
(Also fixed wrong class name in https://github.com/Alexays/Waybar/wiki/Module:-Workspaces)

**Comment by VincentBerthier** [comment]
Hi,
It seems the icons for the workspaces specified in the configuration only work if they’re specified for a workspace ID, not its name.

**Comment by maxwelljens** [comment]
For me, multi-monitor displaying does not work. All the workspaces are shown on both bars in my dual monitor setup. Names also do not work, with Waybar claiming that the argument `{name}` does not exist.

I can open up a separate issue for this, if it sounds interesting.

**Comment by VincentBerthier** [comment]
That’s not me for the unseen ^^
I changed my configuration to use IDs instead of names, it works fine now. I think.

**Comment by maxwelljens** [comment]
@zjeffer I tried many different configurations and the one you presented is one of those I tried. None of the options seem to do anything, except `format`.

```
    "hyprland/workspaces": {
        "format": "{id}",
        "all-outputs": false,
        "persistent_workspaces": {
                "*": 5
            }
    },
```

The error in question is this, if you set `"format": "{name}"`:

```
[2023-08-11 08:54:32.006] [error] hyprland/workspaces: argument not found
[2023-08-11 08:54:32.121] [error] hyprland/workspaces: argument not found
```

It is definitely that element, because if I set `"format": "{id}"` or `"format": "{}"` then it does not show that error. No other errors are present either.

> What do you mean by "unseen workspace"?

I should have said "not visible" instead. If we have three workspaces 1, 2, and 3:

- 1 is visible on left monitor
- 2 is visible on right monitor
- 3 is not visible, whether assigned to left or right monitor
- If I switch between 1 and 2 in Hyprland, then the active workspace in Waybar is always shown as being either 1 or 2, regardless of which workspace is actually active in Hyprland.
- If I switch to 3 in Hyprland, then the active workspace will correctly switch to 3 in Waybar, as expected.

Hope that was clear enough.

**Comment by zjeffer** [comment]
That's so strange, I don't have any of those issues. 

Here's a video of me moving my mouse between two of my monitors: 


https://github.com/Alexays/Waybar/assets/4633209/36a3b3a4-8dbd-4988-bd9b-7c3bae78e13c

It was kind of difficult to record so I only managed to record the workspace regions of my left and middle monitors. I move the mouse and the active workspace changes. Active means a white line appears at the bottom. Non-empty workspaces are a slightly whiter color. At the end of the video, I moved to my third monitor, which makes the other monitors not have an active workspace. 

This seems to work perfectly fine on my system. Can you double-check you are running the latest version?

**Comment by maxwelljens** [comment]
According to `waybar --version`, I am indeed running the latest version 0.9.20.

**Comment by zjeffer** [comment]
Is that the release version or the latest git commit? Because the 0.9.20 release was published a month ago, while my PR with the changes necessary for persistent workspaces was merged two weeks ago (https://github.com/Alexays/Waybar/commit/600653538b778cf9528bb7a230428418b5b9220b)

I use `v0.9.20-64-gb084bf72` from the AUR, which is the latest git version.

**Comment by maxwelljens** [comment]
I believe it is the release version. I use Waybar as distributed in the Fedora 38 repositories.

**Comment by zjeffer** [comment]
Alright, so you'll have to compile Waybar yourself or wait until the next version releases.

@VincentBerthier Can you check your version as well?

**Comment by VincentBerthier** [comment]
0.9.20 too, so… Hmm… Well, I guess this explains that.

**Comment by proudhon** [comment]
I don't know if this is how the module is intended to work but it is not working as expected for me.

It arbitrarily numbers persistent workspaces from left to right not respecting which workspace I've bound to each monitor via hyprland configs.

I don't know if  it is feasible but we should be able to pick which individual workspace to show as persistent instead of specifying the amount of persistent workspaces or automatically show all workspaces bound to the monitor.
![2023-08-14-215217_hyprshot](https://github.com/Alexays/Waybar/assets/94576412/95fe45b7-d9d0-4972-bfec-46b988492c70)

**Comment by zjeffer** [comment]
> I don't know if this is how the module is intended to work

I use it together with [my fork of Duckonaut's split-monitor-workspaces plugin](https://github.com/zjeffer/split-monitor-workspaces). This plugin automatically binds workspaces to monitors, and sets up keybinds so that you can for example do this:

* Automatically bind 5 workspaces to each monitor, based on the formula:
  * `workspace_id = monitor_id * nr_of_workspaces_per_monitor + 1`
  * So, the first workspace on the second monitor (= monitor ID 1) will have `workspace_id` = `1*5+1` = `6`
* Assign keybinds like `Super+4`, to go to the 4th workspace on the currently focused monitor.

The good thing about this approach is that I can use the exact same formula in Waybar, so that the binds the plugin makes correspond to the same workspaces that are shown on each monitor's bar. The plugin assigns workspace 6 to the second monitor, and Waybar also assigns workspace 6 to that monitor. Keybinds work perfectly and without configuration.

If I connect a monitor it has never seen before, it will automatically create 5 persistent workspaces in Waybar and the plugin. All without ever touching Hyprland's config file: no need to set binds. I just tell Waybar and the plugin settings to both add 5 workspaces per monitor, and that's it.

---

The problem with this approach is exactly as you describe: if you manually set your own binds in your hyprland config, the IDs you set in your Hyprland config won't necessarily match Waybar's formula. 

The solution to this is manually setting the persistent_workspaces just like you would in wlr/workspaces (https://github.com/Alexays/Waybar/wiki/Module:-Workspaces#persistent-workspaces). The good news: I also implemented this in hyprland/workspaces :)

Instead of using `"*": 5`, you can manually set which monitors should show which workspaces:

```jsonc
"persistent_workspaces": {
        "1": ["DP-1"], // workspace ID : [array of monitors to show this workspace on, ...]
        "2": ["DP-1"],
        "3": ["DP-1"]
        // and so on, up to 5...
        "6": ["DP-2"],
        // up to 10...
        "11": ["DP-3"],
        // up to 15
}
```

I dislike this approach which is why I implemented the `"*": 5` automatic approach, but I understand if you want to do it manually like above, to have more control over the workspace names.

Let me know if this works for you @proudhon

**Comment by proudhon** [comment]
Thank you so much for the explanation, I couldn't find the manual setup in the wiki and therefore I ended up being confused.

I will try both the approaches and I will give you a feedback asap.

**Comment by proudhon** [comment]
Manual setup works flawlessly, thank you so much.
I will try your own approach asap as it looks more flexible (especially because my docking station tends to mess with  my monitor names very often).

**Comment by zjeffer** [comment]
@mischw That is not yet implemented, I'll see if I can implement that today.

**Comment by zjeffer** [comment]
@mischw I just looked at the Hyprland wiki and the description of those settings implied it should work with the current implementation, so I tested it on my system and it already seems to work fine. 

Are you using the latest release?

**Comment by mischw** [comment]
@zjeffer Oh sorry, I didn't see that there was just a new release. I will try that. Can't get it to compile right now but will try again tomorrow.

**Comment by jr64** [comment]
Fantastic work, thank you very much for implementing!

Is there a way to style workspaces that are currently visible on a monitor differently than those that are not shown on any monitor?

**Comment by zjeffer** [comment]
> Is there a way to style workspaces that are currently visible on a monitor differently than those that are not shown on any monitor?

@jr64 Not yet, I'm thinking of implementing that soon, though

**Comment by khaneliman** [comment]
Just curious if we need to go down this road or request something more similar to the approach we're taking in nixpkgs in discussion https://github.com/NixOS/nixpkgs/pull/251115 about just enabling the experimental flag and using wlr/workspaces as-is. It looks like it functions fine so I'm not sure whether its worth investing effort into this module anymore. @zjeffer  opinions/context?

**Comment by zjeffer** [comment]
@khaneliman Some things that come to mind why I dislike that approach:

* The persistent workspaces implementation is quite limited in wlr/workspaces. 
*  With wlr/workspaces, you have to specify `"<workspace>": [<array of monitors to show workspace on>]`. This is quite cumbersome and doesn't automatically work with keybinds. With my changes in hyprland/workspaces, you can simply specify `"*": 5` to show 5 workspaces on each monitor.
* Persistent workspaces are named automatically based on hyprland monitor IDs, to make sure the names match between waybar and plugins like [split-monitor-workspaces](https://github.com/zjeffer/split-monitor-workspaces). 
* This would need to be ported to wlr/workspaces but it uses hyprland specific stuff (monitor IDs) to accomplish the automatic workspace naming. I actually tried to implement this before hyprland/workspaces existed, see here: https://github.com/Alexays/Waybar/pull/2251. 
* See [this issue](https://github.com/Alexays/Waybar/issues/2243) for more info about why I implemented it this way.

**Comment by khaneliman** [comment]
> @khaneliman Some things that come to mind why I dislike that approach:
> 
>     * The persistent workspaces implementation is quite limited in wlr/workspaces.
> 
>     * With wlr/workspaces, you have to specify `"<workspace>": [<array of monitors to show workspace on>]`. This is quite cumbersome and doesn't automatically work with keybinds. With my changes in hyprland/workspaces, you can simply specify `"*": 5` to show 5 workspaces on each monitor.
> 
>     * Persistent workspaces are named automatically based on hyprland monitor IDs, to make sure the names match between waybar and plugins like [split-monitor-workspaces](https://github.com/zjeffer/split-monitor-workspaces).
> 
>     * This would need to be ported to wlr/workspaces but it uses hyprland specific stuff (monitor IDs) to accomplish the automatic workspace naming. I actually tried to implement this before hyprland/workspaces existed, see here: [[wlr/workspaces] Improve persistent_workspaces by allowing fixed nr of workspaces per monitor #2251](https://github.com/Alexays/Waybar/pull/2251).
> 
>     * See [this issue](https://github.com/Alexays/Waybar/issues/2243) for more info about why I implemented it this way.

The interesting thing is that the wlr implementation supports my use case better of wanting to show specific workspaces on a monitor, albeit in poor experience of declaring it in configuration. I use specific icons/names per monitor and can't seem to get that functioning with hyprland persistent workspaces. 

wlr/workspaces 
![image](https://github.com/Alexays/Waybar/assets/1778670/d4de1822-a3b1-4fb9-8be2-872bff886538)

hyprland/workspaces
![image](https://github.com/Alexays/Waybar/assets/1778670/b9f41bb0-9ab1-4137-b526-a177a30f96d9)

The video workspace is assigned to a different monitor but the perisistent workspace implementation in hyprland shows it on my other monitor.

**Comment by zjeffer** [comment]
> The interesting thing is that the wlr implementation supports my use case better of wanting to show specific workspaces on a monitor, albeit in poor experience of declaring it in configuration. I use specific icons/names per monitor and can't seem to get that functioning with hyprland persistent workspaces.

Your wlr persistent workspace configuration should be able to work the same on both modules, unless I made a mistake while programming it ;)

> The video workspace is assigned to a different monitor but the perisistent workspace implementation in hyprland shows it on my other monitor. 

That sounds more like a bug than an intended feature. Can you create a separate issue for this, with a minimal reproducible example?

**Comment by khaneliman** [comment]
> > The interesting thing is that the wlr implementation supports my use case better of wanting to show specific workspaces on a monitor, albeit in poor experience of declaring it in configuration. I use specific icons/names per monitor and can't seem to get that functioning with hyprland persistent workspaces.
> 
> Your wlr persistent workspace configuration should be able to work the same on both modules, unless I made a mistake while programming it ;)
> 
> > The video workspace is assigned to a different monitor but the perisistent workspace implementation in hyprland shows it on my other monitor.
> 
> That sounds more like a bug than an intended feature. Can you create a separate issue for this, with a minimal reproducible example?

My understanding from the wiki was that I had to just specify a number of workspaces for a monitor and then it just loops through and assigns that many. 
```
{
  "hyprland/workspaces": {
    "all-outputs": false,
    "active-only": "false",
    "format": "{icon}",
    "format-icons": {
      "1": "",
      "2": "",
      "3": "",
      "4": "",
      "5": "",
      "6": "",
      "7": "",
      "8": "󰢹",
      "urgent": "",
      "default": "",
      "persistent": "persistent",
      "empty": "empty"
    },
    "persistent_workspaces": {
      "*": 8,
      "DP-3": 1
    }
  },
```
vs 
```
  "wlr/workspaces": {
    "all-outputs": false,
    "active-only": "false",
    "on-click": "activate",
    "format": "{icon}",
    "format-icons": {
      "1": "",
      "2": "",
      "3": "",
      "4": "",
      "5": "",
      "6": "",
      "7": "",
      "8": "󰢹",
      "urgent": "",
      "default": ""
    },
    "persistent_workspaces": {
      "1": [
        "DP-3"
      ],
      "2": [
        "DP-1"
      ],
      "3": [
        "DP-1"
      ],
      "4": [
        "DP-1"
      ],
      "5": [
        "DP-1"
      ],
      "6": [
        "DP-1"
      ],
      "7": [
        "DP-1"
      ],
      "8": [
        "DP-1"
      ]
    }
  }
```

**Comment by zjeffer** [comment]
Yeah the wiki should be updated with that info and more examples but I've been kind of busy with other stuff lately and didn't feel like it :)

**Comment by zjeffer** [comment]
@khaneliman 

> hyprland/workspaces ideal but probably hard to support

That actually looks pretty easy to implement, I'll do it if I find some time.

**Comment by khaneliman** [comment]
> @khaneliman
> 
> > hyprland/workspaces ideal but probably hard to support
> 
> That actually looks pretty easy to implement, I'll do it if I find some time.

I guess just check if it's an array within it instead of a single number to process differently.

**Comment by zjeffer** [comment]
> I guess just check if it's an array within it instead of a single number to process differently 

Indeed.

**Comment by MightyPlaza** [comment]
special workspaces can now opened empty in hyprland
so persistent special workpsaces can exist and behaviour should be tested

**Comment by zjeffer** [comment]
>  so persistent special workspaces can exist and behavior should be tested

Current behaviour with this config:

```jsonc
"hyprland/workspaces": {
    "format": "{icon}",
    "format-icons": {
      "urgent": "",
      "active": "",
      "default": "",
      "persistent": "",
      "special": "s"
    },
    "on-scroll-up": "hyprctl dispatch workspace r-1",
    "on-scroll-down": "hyprctl dispatch workspace r+1",
    "all-outputs": false,
    "show-special": true,
    "persistent_workspaces": {
      "*": 5,
      "special": [
        "eDP-1",
        "DP-2",
        "HDMI-A-1"
      ]
    }
  },
```
seems to show the workspace persistently on all my monitors, and has the correct css class both when empty and not empty.

There seems to be no active class when the special workspace is toggled. I would assume the user wants visual feedback when the special workspace is active, right? Should the currently active workspace keep its active class or should it become non-active. I think it should keep the active class.

There's also some weird behaviour with named special workspaces, I'll have to investigate further.

**Comment by MightyPlaza** [comment]
there isn't a IPC event for when a special workspace is opened
you can find out by calling `hyprcctl monitors`
feel free to create a feature request, but for now the best you can do is start preparing the code for when it gets implemented

maybe add a format for special-active

> There's also some weird behaviour with named special workspaces, I'll have to investigate further.
?

**Comment by cyrinux** [comment]
Hi,

Do I miss something? I can't get working https://github.com/hyprland-community/hyprland-autoname-workspaces with this new module. What is the way to keep` the same format and behavior as `wlr/workspaces` ?

* And also, `wlr/workspaces` didn't work anymore, but I can't find error, nothing but no workspaces anymore at all with this legacy module.
* I would like to be able to still use hyprland-autoname-workspaces to control the content of my workspaces in waybar.

Thanks

**Comment by khaneliman** [comment]
> * And also, `wlr/workspaces` didn't work anymore for me, but I can't find relevant error, nothing but no workspaces anymore at all with this legacy(?) module.
> 
> 

The protocol that supported wlr/workspaces module in Hyprland was removed from Hyprland recently.

**Comment by khaneliman** [comment]
>     * I would like to be able to still use hyprland-autoname-workspaces to control the content of my workspaces in waybar.

I've never used that program before, but I can try and take a look at what the workspaces module needs to make it work.

**Comment by cyrinux** [comment]
> > ```
> > * I would like to be able to still use hyprland-autoname-workspaces to control the content of my workspaces in waybar.
> > ```
> 
> I've never used that program before, but I can try and take a look at what the workspaces module needs to make it work.

I would be happy if you can do this please 👍🏻

**Comment by cyrinux** [comment]
> > * And also, `wlr/workspaces` didn't work anymore for me, but I can't find relevant error, nothing but no workspaces anymore at all with this legacy(?) module.
> 
> The protocol that supported wlr/workspaces module in Hyprland was removed from Hyprland recently.

https://github.com/hyprwm/Hyprland/commit/bb0933437f4602080a66cc877fee36dce8dcb3ff OMG i'm so sad.

**Comment by MightyPlaza** [comment]
new PR for "activespecial" IPC event is up so special workspace integration should be fully doable now

**Comment by zjeffer** [comment]
> new PR for "activespecial" IPC event is up so special workspace integration should be fully doable now

Alright, I'll probably do this next weekend because I'll be too tired after work during the week... Unless someone else wants to implement it? ;)

**Comment by niksingh710** [comment]
for me the persistent workspace icon is shown every where i wanted to modify the workspaces somehow so that occupied have different icons then non occupied workspace.
it was possible int `wlr/workspace` as then occupied used to have only the `default: icons` and all unoccupied used to have `persistent: icon` but in hyprland/workspace `persistent` icons override the default one's

**Comment by zjeffer** [comment]
@niksingh710 If I understand your request correctly, you should be able to get the behaviour you want by using the "empty" icon class. 

Occupied persistent workspaces will have the persistent icon, while non-occopied persistent workspaces will have the empty icon. If you don't define the empty icon in your config, it will fall back to the persistent icon. Hope this helps.

**Comment by MichaelPetersen22** [comment]
Is it possible to add a visible workspaces class?

**Comment by niksingh710** [comment]
> Occupied persistent workspaces will have the persistent icon, while non-occopied persistent workspaces will have the empty icon. If you don't define the empty icon in your config, it will fall back to the persistent icon. Hope this helps.

Exactly what I wanted.
I guess it wasn't on the wiki earlier.

**Comment by cyrinux** [comment]
> > ```
> > * I would like to be able to still use hyprland-autoname-workspaces to control the content of my workspaces in waybar.
> > ```
> 
> I've never used that program before, but I can try and take a look at what the workspaces module needs to make it work.

I understand we maybe need a "mode" in this module, that will "just" fetch for example every second, or on events a fetch of workspaces name. Maybe react on the same events as hyprland-autoname-workspaces to be sure in this mode.

**Comment by cyrinux** [comment]
The `renameworkspace` was added https://github.com/hyprwm/Hyprland/commit/2d100bf57e0c0184d307bd4173b1a8d9fc7130e1#diff-3c089c6c6b23a85ab5b4cffb002e8880ad069c8a8b3462f7aafa9e5345c0333d

**Comment by tumon-art** [comment]
`#workspaces button.urgent `

isn't working anymore!

**Comment by khaneliman** [comment]
> #workspaces button.urgent 
>
> isn't working anymore!

In what way? I daily drive the last commit and it works for me.

**Comment by tumon-art** [comment]
> > #workspaces button.urgent
> > isn't working anymore!
> 
> In what way? I daily drive the last commit and it works for me. My example of styling in other issue: [#2469 (comment)](https://github.com/Alexays/Waybar/issues/2469#issuecomment-1707603189)

It works with "wlr/workspaces" module in **hyprland 28**
I tired with "hyprland/workspaces" module in **hyparland 28** but it doesn't work.

I'm using **hyprland 29.1** now and "hyprland/workspaces" module.
It doesn't work for me.

**Comment by zjeffer** [comment]
@tumon-art But which version of waybar are you using? The urgent class was only added pretty recently.

**Comment by tumon-art** [comment]
> @tumon-art But which version of waybar are you using? The urgent class was only added pretty recently.

**0.9.22**

**Comment by khaneliman** [comment]
> > @tumon-art But which version of waybar are you using? The urgent class was only added pretty recently.
> 
> **0.9.22**
> 
> 

Urgent isn't in that release.

**Comment by NotAShelf** [comment]
There seems to be an issue with active-only after https://github.com/Alexays/Waybar/pull/2408 has been implemented - you will see all active workspaces for the fraction of a second while switching from an active workspace to an empty one.

CC @zjeffer

**Comment by khaneliman** [comment]
> Sorry to necro the split-monitor discussion, but is there an efficient way to rename split-workspaces to their relative number?
> 
> For example: There are 5 workspaces per monitor. Thus, workspace id "6" corresponds to workspace 1 on the second monitor. Instead of showing this as "6" in the second monitor's waybar, I would like an option to have it shown as "1".
> 
> In short, the ability to set a number `n` such that the displayed workspace names cycle every `n` workspaces.

Does format-icon not work? Just map 6-10 as 1-5 again for display.

**Comment by droc12345** [comment]
Not sure if this the place to ask, but what would it take to incorporate something like wlr-taskbar into hyprland-workspaces?
If not feasible (I haven't checked for complexity) then would it be possible to either have a tooltip & tooltip-format added,
that way I could run a script to get info from hyprland on a hover event. 
I suppose on-click might also work.

Anyway, just asking, and thanks to all doing the work to make waybar/* possible

**Comment by constantitus** [comment]
A feature which I really need would be the option to have a different format-icon for "ocupied" workspaces.
There are currently format-icons available for default, active and urgent workspaces. Is there any way this could be implemented ?

**Comment by niksingh710** [comment]
```json
  "hyprland/workspaces": {
    "format": "{icon}",
    "on-click": "activate",
    "persistent_workspaces": {
      "1": [],
      "2": [],
      "3": [],
      "4": [],
      "5": [],
      "6": [],
      "7": [],
      "8": [],
      "9": [],
      "10": []
    },
    "format-icons": {
      "active": " 󰮯 ",
      "default": " ",
      "empty": "󰧵 ",
    }
  }```

This is my config.
earlier it used to display all workspaces (1 to 10) with correct icons
now this same config does not work as expected.
it only shows the occupied workspaces.

**Comment by Diaoul** [comment]
Hyprland has native support for persistent workspaces now (see https://github.com/hyprwm/Hyprland/pull/3530)
This is not needed anymore.

**Comment by niksingh710** [comment]
> Hyprland has native support for persistent workspaces now (see [hyprwm/Hyprland#3530](https://github.com/hyprwm/Hyprland/pull/3530)) This is not needed anymore.

Thanks for mentioning mate seems like have to wait for a while to get it in release.

**Comment by Diaoul** [comment]
I'm using the git version until the next releases comes out. I think you can do the same :shrug:

Worth noting that waybar behaves weirdly when moving workspaces around across monitors, reloading configuration solves the issue: `killall -SIGUSR2 waybar`

**Comment by zjeffer** [comment]
> Hyprland has native support for persistent workspaces now (see [hyprwm/Hyprland#3530](https://github.com/hyprwm/Hyprland/pull/3530)) This is not needed anymore.

Great! Once [this branch](https://github.com/hyprwm/Hyprland/compare/main...hariseldon78:Hyprland:persistent-workspaces) is merged, we should be able to query for persistent workspaces and we can then dynamically fill the module with the persistent workspaces, without having to manually specify them in the config :tada:

**Comment by hariseldon78** [comment]
> @hariseldon78 This would definitely be very useful for us!

Ok great to hear that! I'll get to it asap.

**Comment by hariseldon78** [comment]
https://github.com/hyprwm/Hyprland/pull/3630 it's merged guys, let me know if you need any help using the feature

**Comment by niksingh710** [comment]
> Hyprland has native support for persistent workspaces now (see [hyprwm/Hyprland#3530](https://github.com/hyprwm/Hyprland/pull/3530)) This is not needed anymore.

This is my hyprland workspaces

```json
[{
    "workspaceString": "1",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "2",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "3",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "4",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "5",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "6",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "7",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "8",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "9",
    "monitor": "DP-1",
    "default": true,
    "persistent": true
},{
    "workspaceString": "10",
    "monitor": "eDP-1",
    "default": true,
    "persistent": true
}]
```


This is my waybar Config
```json
  "hyprland/workspaces": {
    "format": "{icon}",
    "on-click": "activate",
    "all-outputs": false,
    "format-icons": {
      "active": " 󰮯 ",
      "default": " ",
      "empty": "󰧵 ",
      "persistent": "󰧵 ",
    }
  },
```

only occupied workspaces are shown.

**Comment by niksingh710** [comment]
one thing i observed is if i switch my focus to an empty workspace then it is getting shown in waybar but then also it is not using `empty` icon defined in config. 

and if i logout login back again i need to focus each workspace once then only it shows up.

**Comment by khaneliman** [comment]
> > Hyprland has native support for persistent workspaces now (see [hyprwm/Hyprland#3530](https://github.com/hyprwm/Hyprland/pull/3530)) This is not needed anymore.
> 
> This is my hyprland workspaces
> 
> ```json
> [{
>     "workspaceString": "1",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "2",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "3",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "4",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "5",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "6",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "7",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "8",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "9",
>     "monitor": "DP-1",
>     "default": true,
>     "persistent": true
> },{
>     "workspaceString": "10",
>     "monitor": "eDP-1",
>     "default": true,
>     "persistent": true
> }]
> ```
> 
> This is my waybar Config
> 
> ```json
>   "hyprland/workspaces": {
>     "format": "{icon}",
>     "on-click": "activate",
>     "all-outputs": false,
>     "format-icons": {
>       "active": " 󰮯 ",
>       "default": " ",
>       "empty": "󰧵 ",
>       "persistent": "󰧵 ",
>     }
>   },
> ```
> 
> only occupied workspaces are shown.

Looks like you didn’t set up persistent workspaces in waybar.

**Comment by hariseldon78** [comment]
@niksingh710  maybe you are already doing it, but those shown are just the workspace rules, they act on the workspaces **when they are present**, so you always need 2 commands, one to set the workspacerule and one to create the workspace. 

That said, i also found some inconsistencies with how waybar react when i create a new persistent workspace, sometimes it is shown many times, sometimes it is shown with the wrong css style (occupied instead of empty or viceversa). My workaround for now is to restart the waybar in the same script, and it works. Here is my script, it creates a workspace asking the name and then it sorts them alphabetically:

[hyperland-create-workspace.py](https://gist.github.com/hariseldon78/d0b91e584fa97e3efcbd280cf1703f5f#file-hyprland-create-workspace-py)

**Comment by Diaoul** [comment]
Here is [my script](https://gist.github.com/Diaoul/4a33dfec9d80763a7f74ce8086bf5237#file-monitors-sh) maintaining 10 workspaces on multiple monitors.

**Comment by zjeffer** [comment]
@niksingh710 FYI: Hyprland's new persistent workspace rule (implemented in Hyprland by @hariseldon78) is not yet supported in Waybar. I'm working on it in this PR: https://github.com/Alexays/Waybar/pull/2603, but I haven't had much time to work on it lately. For now, you should specify your persistent workspaces in your waybar config, see the wiki for an example.

**Comment by niksingh710** [comment]
> @niksingh710 FYI: Hyprland's new persistent workspace rule (implemented in Hyprland by @hariseldon78) is not yet supported in Waybar. I'm working on it in this PR: #2603, but I haven't had much time to work on it lately. For now, you should specify your persistent workspaces in your waybar config, see the wiki for an example.

Added persistent workspace.
I think hyprland persistent workspace making them identify as non-empty as now `e+1` and `r+1` behave in same manner.
I have asked that in the pr discussion https://github.com/hyprwm/Hyprland/pull/3530

**Comment by zjeffer** [comment]
> after the rename event gets registered and handled i tried to use renaming more and more. But currently after renaming the order of workspaces stays the same although i configured 'sort-by': 'name'.
>
> Is there any chance i can trigger the sort from outside or this behaviour gets added?

Looks like we're already listening to workspace rename events, maybe we should add a call to `sort_workspaces()` at the end of this function: https://github.com/Alexays/Waybar/blob/0a747ddafc10ace2a2d8588ba673b9c2015d8446/src/modules/hyprland/workspaces.cpp#L329-L342

**Comment by zjeffer** [comment]
Hi all, I finally found time to finish my PR to use Hyprland's persistent workspace rules instead of Waybar's "persistent-workspaces" object, it's ready to test and review: https://github.com/Alexays/Waybar/pull/2603

Remember to update Hyprland to be able to listen to the `configreloaded` event.

**Comment by pkreuzt** [comment]
Currently module cycles with scroll through workspaces back to beginning when the last one is reached. Could you add an option to configure/avoid this? Something like `disable-scroll-wraparound` option in Sway/workspaces module.

**Comment by zjeffer** [comment]
@pkreuzt I added this to my "hyprland/workspaces" object in my waybar config:

```
"on-scroll-up": "hyprctl dispatch workspace r-1",                                                                                                                                
"on-scroll-down": "hyprctl dispatch workspace r+1"
```

**Comment by pkreuzt** [comment]
That does something not very desirable for me. Could work as a temporary workaround, but the issue is still there.

**Comment by zjeffer** [comment]
@pkreuzt Can you explain what you mean by 'something not very desirable'?
What behavior do you want to see instead? Do you want the scroll to not do anything at all when you reach the start/end?

**Comment by pkreuzt** [comment]
Previous config goes on creating new workspaces beyond existing ones. Or should I say, makes empty workspaces visible on scrolling when I have chosen to show only occupied ones. Some people would maybe like this, but not everyone.

There should be a way to scroll only through busy workspaces (or fixed ones) but without looping. For example, if I have 3 workspaces visible on module,  scrolling over it from 1 should move to 2 and 3 at most, and stay there without looping back to 1.

**Comment by zjeffer** [comment]
I see. You said: `Currently module cycles with scroll through workspaces back to beginning when the last one is reached`. Can you share your config that causes this? I can't reproduce that with a minimal config (scrolling doesn't do anything)

**Comment by pkreuzt** [comment]
Module config is like this:

```
	"hyprland/workspaces": {
                 "on-scroll-up": "hyprctl dispatch workspace m-1",
                 "on-scroll-down": "hyprctl dispatch workspace m+1",
		 "active-only": false,
		 "all-outputs": false,
	    },

```

**Comment by zjeffer** [comment]
Got it. Can you create a separate GitHub issue with the explanation of your feature request? As far as I know it's not yet possible to do what you want.

**Comment by Alexays** [comment]
Thanks @zjeffer for driving this tracking issue — it did its job. Nearly everything on the list has shipped: urgent/focused/visible/persistent classes, active-only, per-output control, on-click/on-scroll, and the refactor to Hyprland's persistent-workspace rule (#2603). The hyprland/workspaces module is mature now.

The couple of remaining items already live as their own issues (#2484, #2555) and the special-workspace integration can be filed/followed separately, so this umbrella issue is stale and no longer the best place to track them.

Closing as superseded by the individual issues — please open focused issues for any specific remaining request and we'll pick them up there. 🙏
